src/v5_hal/firmware/src/nodes/LiftNode.cpp

```diff
@@ -49,49 +52,125 @@ void LiftNode::setLiftVelocity(int velocity) {
+        case UP_FOR_RINGS: 
+            setLiftPosition(900); // UNTESTED
+        break;
+        
+        case FULLY_UP:
+            setLiftPosition(0); // UNTESTED
+        break;
+
+        default:
+        break;
+    }
+
     int positionBoundUpper = getPosition() + m_tolerance;
     int positionBoundLower = getPosition() - m_tolerance;
     if(positionBoundLower < m_target_position && m_target_position < positionBoundUpper) {
-        m_lift_state = HOLDING;
+        setLiftVelocity(0);
     } else {
-        m_lift_state = UPDATING;
+        m_setLiftPID();
     }
 }
 
-void LiftNode::teleopPeriodic() {
-    if (m_controller->getController()->get_digital(pros::E_CONTROLLER_DIGITAL_R1) && 
-        !m_controller->getController()->get_digital(pros::E_CONTROLLER_DIGITAL_R2)) {
-        setLiftVoltage(MAX_MOTOR_VOLTAGE); // using voltage here cause we don't know max motor velocity for sure
-    } else if (m_controller->getController()->get_digital(pros::E_CONTROLLER_DIGITAL_R2) && 
-        !m_controller->getController()->get_digital(pros::E_CONTROLLER_DIGITAL_R1)) {
-        setLiftVoltage(-MAX_MOTOR_VOLTAGE);
-    } else {
-        setLiftVelocity(0);
+/**
+ * Only called in teleopPeriodic()
+ * Does not currently have a way to switch to the free moving state
+ * */
+void LiftNode::m_updateLiftStateTeleop() { 
+    bool moveUp = false;
+    bool moveDown = false;
+    
+    // this logic is the exact same as ClawNode I wonder 
+    // how we could combine the two
+    bool upButtonCurrentState = m_controller->getController()->get_digital(m_upButton);
+    bool downButtonCurrentState = m_controller->getController()->get_digital(m_downButton);
+
+    // there feels like there is a better way to do the state logic when a button is pressed
+	if ((upButtonCurrentState == 1 && m_upButtonPreivousState == 0) &&
+            !(downButtonCurrentState == 1 && m_downButtonPreviousState == 0)) {
+        moveUp = true;
+    }
+
+    if ((downButtonCurrentState == 1 && m_downButtonPreviousState == 0) &&
+            !(upButtonCurrentState == 1 && m_upButtonPreivousState == 0)) {
+        moveDown = true;
     }
-};
 
-void LiftNode::autonPeriodic() {
-    updateLiftState();
+	m_upButtonPreivousState = upButtonCurrentState;
+    m_downButtonPreviousState = downButtonCurrentState;
     
-    switch (m_lift_state) {
-        case UPDATING:
-            m_setLiftPID();
+    switch (m_lift_state) { // write this logic
+        case DOWN:
+            if (moveUp) {
+                m_lift_state = UP_FOR_RINGS;
+            }
         break;
-        case HOLDING:
-            setLiftVelocity(0);
+        
+        case UP_FOR_RINGS:
+            if (moveUp) {
+                m_lift_state = FULLY_UP;
+            } else if (moveDown) {
+                m_lift_state = DOWN;
+            }
+        break;
+        
+        case FULLY_UP:
+            if (moveDown) {
+                m_lift_state = UP_FOR_RINGS;
+            }
         break;
     }
-};
+}

```

Comment:
I'd want to talk with Brian on this one - in my opinion, it's still important to have some means to manually control the lift. With the goal lift that we have it's probably not as important, but still something to think about.

I'd keep going down the exact route you're on with the logic behind controlling the lift, as I think this is the right way to do it. I'd be interested in talking to Brian to see what he thinks about manual control vs. all automated

Reply:
ex. if a sensor breaks, or something gets damaged, we could lose full control of the lift versus having manual control

Reply:
That is my main concern as well. I will probably try to implement